utils/httputil/server.go

```diff
@@ -0,0 +1,50 @@
+package httputil
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+)
+
+type grace interface {
+	Shutdown(context.Context) error
+}
+
+func GracefulListenAndServe(ctx context.Context, server *http.Server, shutdownTimeout ...time.Duration) error {
+	return gracefulFunc(ctx, server, server.ListenAndServe, shutdownTimeout...)
+}
+
+func GracefulServe(ctx context.Context, server *http.Server, l net.Listener, shutdownTimeout ...time.Duration) error {
+	fn := func() error {
+		return server.Serve(l)
+	}
+	return gracefulFunc(ctx, server, fn, shutdownTimeout...)
+}
+
+func gracefulFunc(ctx context.Context, g grace, fn func() error, shutdownTimeout ...time.Duration) error {
```

Comment:
How come `shutdownTimeout` is a variadic parameter?

Reply:
If it's to make it optional, shouldn't you check whether it is `> 0` anyway? Seems bad to pass a 0 there, it would just be doomed to fail :sweat_smile:

Reply:
The reason a made variadic is that in some cases is not possible to figure out what is the correct `shutdownTimeout`. 

You might have a lot of application-specific logic that dictates the timeout per handler, and figuring out the correct value on the top level is impossible.

In those cases, the responsibility is on the handler to use context and timeout properly, and a Shutdown should have no timeout.

An improper shutdown will report all running go routines - so it can be examined.

Reply:
I think I should make `0` zero helpful value here, and cancel the server instead of calling shutdown

Reply:
Sounds good :+1: